awx/api/generics.py

```diff
@@ -153,13 +154,21 @@ def initialize_request(self, request, *args, **kwargs):
         Store the Django REST Framework Request object as an attribute on the
         normal Django request, store time the request started.
         """
+        is_trusted_proxy = False
+
         self.time_started = time.time()
         if getattr(settings, 'SQL_DEBUG', False):
             self.queries_before = len(connection.queries)
 
+        if 'HTTP_X_TRUSTED_PROXY' in request.META:
+            if validate_x_trusted_proxy_header(request.META['HTTP_X_TRUSTED_PROXY']):
+                is_trusted_proxy = True
+            else:
+                logger.warning("Request appeared to be a trusted upstream proxy but failed to provide a matching shared secret.")
+
         # If there are any custom headers in REMOTE_HOST_HEADERS, make sure
         # they respect the allowed proxy list
-        if all(
+        if not is_trusted_proxy and all(
```

Comment:
I see you have tests below but I am reading this as if:
    * I am not trusted and
    * I have PROXY_IP_ALLOWED_LIST and
    * My REMOTE_ADDRS it not in PROXY_IP_ALLOWED_LIST and
    * My REMOTE_HOST is not in PROXY_IP_ALLOWED_LIST
then remove the headers.

Wouldn't the not_trusted just let you decide if you can look through more headers than just REMOTE_ADDRS and REMOTE_HOST and not whether or not we should remove headers?

Reply:
Yes. Making the necessary changes.